### ISSUE TITLE
AV1 compille error

### ISSUE THREAD

**Issue by jojo61** [haasn/libplacebo#69]
error: glslang failed: ERROR: 0:59: '' :  syntax error, unexpected COMMA
ERROR: 1 compilation errors.  No code generated.

error: [  1] #version 450
error: [  2] #extension GL_ARB_compute_shader : enable
error: [  3] layout (local_size_x = 32, local_size_y = 32) in;
error: [  4] layout(std430, push_constant) uniform PushC {
error: [  5]     layout(offset=0) vec2 _out_scale_12_0;
error: [  6]     layout(offset=8) vec2 _p0_13_0;
error: [  7]     layout(offset=16) vec2 _p1_14_0;
error: [  8]     layout(offset=24) vec2 _p2_15_0;
error: [  9]     layout(offset=32) vec2 _p3_16_0;
error: [ 10]     layout(offset=40) ivec2 _base_18_0;
error: [ 11] };
error: [ 12] layout(binding=0) uniform sampler2D _src_tex_0_0;
error: [ 13] layout(binding=1) uniform sampler1D _weights_3_0;
error: [ 14] layout(binding=2) uniform sampler1D _weights_6_0;
error: [ 15] layout(binding=3) uniform sampler1D _weights_9_0;
error: [ 16] layout(std430, binding=4) readonly buffer _AV1GrainBuf_11_0 {
error: [ 17]     layout(offset=0) float grain_y[4096];
error: [ 18]     layout(offset=16384) float grain_cb[4096];
error: [ 19]     layout(offset=32768) float grain_cr[4096];
error: [ 20]     layout(offset=49152) uint offsets[960];
error: [ 21] };
error: [ 22] layout(binding=5, r8) writeonly uniform image2D _out_image_17_0;
error: [ 23] #define frag_pos(id) (vec2(id) + vec2(0.5)) 
error: [ 24] #define frag_map(id) (_out_scale_12_0 * frag_pos(id))    
error: [ 25] #define gl_FragCoord vec4(frag_pos(gl_GlobalInvocationID), 0.0, 1.0) 
error: [ 26] #define _tex_coord_1_0_map(id) (mix(mix(_p0_13_0, _p1_14_0, frag_map(id).x),      mix(_p2_15_0, _p3_16_0, frag_map(id).x), frag_map(id).y))
error: [ 27] #define _tex_coord_1_0 (_tex_coord_1_0_map(gl_GlobalInvocationID))
error: [ 28] #define _LUT_POS_4_0(x) mix(0.001953, 0.998047, (x)) 
error: [ 29] #define _lut_2_0(pos) (texture(_weights_3_0, float(\
error: [ 30]     _LUT_POS_4_0((pos).x)\
error: [ 31]   )).x)
error: [ 32] #define _LUT_POS_7_0(x) mix(0.001953, 0.998047, (x)) 
error: [ 33] #define _lut_5_0(pos) (texture(_weights_6_0, float(\
error: [ 34]     _LUT_POS_7_0((pos).x)\
error: [ 35]   )).x)
error: [ 36] #define _LUT_POS_10_0(x) mix(0.001953, 0.998047, (x)) 
error: [ 37] #define _lut_8_0(pos) (texture(_weights_9_0, float(\
error: [ 38]     _LUT_POS_10_0((pos).x)\
error: [ 39]   )).x)
error: [ 40] void _main_19_0() {
error: [ 41] // pl_shader_sample_direct          
error: [ 42] vec4 color = vec4(1.000000) * texture(_src_tex_0_0, _tex_coord_1_0); 
error: [ 43] // pl_shader_av1_grain 
error: [ 44] {                      
error: [ 45] uvec2 offset;          
error: [ 46] uvec2 pos;             
error: [ 47] float val;             
error: [ 48] float grain;           
error: [ 49] uvec2 block_id  = gl_WorkGroupID.xy;        
error: [ 50] uvec2 local_id  = gl_LocalInvocationID.xy;  
error: [ 51] uvec2 global_id = gl_GlobalInvocationID.xy; 
error: [ 52] uint data = offsets[block_id.y * 40 + block_id.x]; 
error: [ 53] offset = 2u * uvec2((data >> 4) & 0xFu,   
error: [ 54]                      (data >> 0) & 0xFu);  
error: [ 55] pos = offset + local_id.xy + uvec2(0, 0); 
error: [ 56] val = grain_y[ pos.y * 64 + pos.x ];       
error: [ 57] grain = val; 
error: [ 58] color[0] += _lut_2_0(color[0] * 1.000000) * grain; 
error: [ 59] color[0] = clamp(color[0], , );    
error: [ 60] } 
error: [ 61] ivec2 dir = ivec2(1, 1);

**Comment by haasn** [comment]
I'm gonna need help reproducing this. Works fine on my end?

**Comment by jojo61** [comment]
I updated today from libplacebo.31 to libplacebo.34 and then this happens.
I did not change anything in my application. I compile libplacebo myself. 
Here is the build output:

The Meson build system
Version: 0.54.0
Source dir: /usr/src/libplacebo-master
Build dir: /usr/src/libplacebo-master/build
Build type: native build
Project name: libplacebo
Project version: 1.34.0
C compiler for the host machine: cc (gcc 7.5.0 "cc (SUSE Linux) 7.5.0")
C linker for the host machine: cc ld.bfd 15.1
C++ compiler for the host machine: c++ (gcc 7.5.0 "c++ (SUSE Linux) 7.5.0")
C++ linker for the host machine: c++ ld.bfd 15.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program git found: YES (/usr/bin/git)
Compiler for C supports arguments -Wincompatible-pointer-types: YES 
Compiler for C supports link arguments -Wl,--exclude-libs=ALL: YES 
meson.build:67: WARNING: Consider using the built-in warning_level option instead of using "-Wall".
meson.build:68: WARNING: Consider using the built-in warning_level option instead of using "-Wall".
meson.build:68: WARNING: Consider using the built-in option for language standard version instead of using "-std=c++11".
Library m found: YES

|Executing subproject xtalloc method meson 
|
|Project name: xtalloc
|Project version: undefined
|C compiler for the host machine: cc (gcc 7.5.0 "cc (SUSE Linux) 7.5.0")
|C linker for the host machine: cc ld.bfd 15.1
|Run-time dependency threads found: YES
|Build targets in project: 1
|Subproject xtalloc finished.


|Executing subproject bstr method meson 
|
|Project name: bstr
|Project version: undefined
|C compiler for the host machine: cc (gcc 7.5.0 "cc (SUSE Linux) 7.5.0")
|C linker for the host machine: cc ld.bfd 15.1
|Build targets in project: 2
|Subproject bstr finished.

Found pkg-config: /usr/bin/pkg-config (0.29.2)
Run-time dependency vulkan found: YES 1.0.65
Run-time dependency epoxy found: YES 1.4.3
Run-time dependency x11 found: YES 1.6.5
Library shaderc_shared found: NO
Library shaderc_combined found: NO
Library glslang found: YES
Library HLSL found: YES
Library OGLCompiler found: YES
Library OSDependent found: YES
Library SPIRV found: YES
Library SPVRemapper found: YES
Library SPIRV-Tools found: NO
Library SPIRV-Tools-opt found: NO
Fetching value of define "GLSLANG_PATCH_LEVEL" : 2933 
Library pthread found: YES
Run-time dependency lcms2 found: YES 2.9
Configuring config.h using configuration
Configuring config_internal.h using configuration
Dependency threads found: YES unknown (cached)
Build targets in project: 3

libplacebo 1.34.0

  Subprojects
       bstr: YES
    xtalloc: YES

Found ninja-1.8.2 at /usr/bin/ninja
ninja: Entering directory `./build'
[40/40] Linking target src/libplacebo.so.34

**Comment by haasn** [comment]
I'm using the same version and still can't reproduce this. What especially confuses me is that it seems to be some fundamental deficiency in the ability to printf GLSL code, the generated code doesn't just contain wrong literals - it contains no literals at all. It's the type of issue I'd more assume is related to something like stack corruption.

**Comment by haasn** [comment]
I fixed some potential UB in av1.c. Can you see if master fixes this issue?

**Comment by jojo61** [comment]
I found the problem and it was my fault.
I forgot to update the header files in /usr/include/libplacebo from the newer version. So the structures did not fit. Sorry for the confusion.

**Comment by haasn** [comment]
Makes sense. Shouldn't the runtime check for `PL_API_VER` have caught this, though? (Unless you included the correct generated `config.h` but failed to include the rest of the headers)

**Comment by jojo61** [comment]
Well I have problems to understand your concept with config.h.  It will be included with #include "config.h" and therefore is needed in my local directory. But the other include files are included from /usr/include/libplacebo. So I need to copy the config.h and the other headers to different destinations in my filesystem. Why is it so ?  Could we not copy the config.h as part of the other headers and include it with #include <libplacebo/config.h> ?

**Comment by haasn** [comment]
`config.h` should be installed alongside the other headers into `/usr/libplacebo`. It should not be in your local directory.

I'm not sure why we include `"config.h"` instead of `<libplacebo/config.h>` but it may have been a work-around for some annoying path-related issue, the intent certainly isn't for users to copy this config.h around.

**Comment by haasn** [comment]
> I'm not sure why we include `"config.h"` instead of `<libplacebo/config.h>` but it may have been a work-around for some annoying path-related issue, the intent certainly isn't for users to copy this config.h around.

I tried it again and the issue is that during the compile process, meson internally templates this out to `meson-private/src/config.h` rather than `meson-private/src/libplacebo/config.h` or any other path containing `libplacebo`. This is enforced quite explicitly, i.e. it raises an error if you try specifying a destination path containing a subdirectory; and it also ignores the source path's directory when templating.

So tl;dr we have to include it as `"config.h"` to make the build process of libplacebo itself work. Suggestions pending.(tm)

**Comment by haasn** [comment]
> Suggestions pending.(tm)

Found an easy way to work around this. Pushed as 2101311.

**Comment by jojo61** [comment]
Perfect. Works like a charm. Thanks
